sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -385,12 +407,16 @@ object TypeCoercion {
     }
 
     /** Given a plan, add an extra project on top to widen some columns' data types. */
-    private def widenTypes(plan: LogicalPlan, targetTypes: Seq[DataType]): LogicalPlan = {
-      val casted = plan.output.zip(targetTypes).map {
-        case (e, dt) if e.dataType != dt => Alias(Cast(e, dt), e.name)()
-        case (e, _) => e
-      }
-      Project(casted, plan)
+    private def widenTypes(plan: LogicalPlan, targetTypes: Seq[DataType])
+      : (LogicalPlan, Seq[(ExprId, (ExprId, DataType))]) = {
+      val (casted, newExprIds) = plan.output.zip(targetTypes).map {
+        case (e, dt) if e.dataType != dt =>
+          val alias = Alias(Cast(e, dt), e.name)()
+          (alias, Some(e.exprId -> (alias.exprId, dt)))
```

Comment:
Can we just store the attribute of the `Alias`, i.e., `alias.toAttribute`?

Reply:
Ah, I see. I'll update.